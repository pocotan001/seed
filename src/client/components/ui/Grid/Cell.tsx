import { CSSProperties } from "react";
import styled from "styled-components";

interface CellProps {
  children: React.ReactNode;
  col?: number;
  row?: number;
  justify?: CSSProperties["justifySelf"];
  align?: CSSProperties["alignSelf"];
  order?: CSSProperties["order"];
}

const Cell = styled("div")<CellProps>`
  ${({ col }) => col && `grid-column-end: span ${col}`};
  ${({ row }) => row && `grid-row-end: span ${row}`};
  ${({ justify }) => justify && `justify-self: ${justify}`};
  ${({ align }) => align && `align-self: ${align}`};
  ${({ order }) => order && `order: ${order}`};
`;

Cell.defaultProps = {
  col: 1
};

export default Cell;
