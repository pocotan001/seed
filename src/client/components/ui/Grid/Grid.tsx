import { CSSProperties } from "react";
import styled from "styled-components";
import margin, { MarginProps } from "../../styles/extends/margin";
import { px } from "../../styles/utils";

interface OwnProps {
  children: React.ReactNode;
  inline?: boolean;
  cols?: number | CSSProperties["gridTemplateColumns"];
  rows?: number | CSSProperties["gridTemplateRows"];
  flow?: CSSProperties["gridAutoFlow"];
  gap?: CSSProperties["gridGap"];
  justify?: CSSProperties["justifyContent"];
  align?: CSSProperties["alignItems"];
}

type GridProps = OwnProps & MarginProps;

const buildColumns = (value: number | string) =>
  typeof value === "number" ? `repeat(${value}, 1fr)` : value;
const buildRows = buildColumns;

const Grid = styled("div")<GridProps>`
  display: ${({ inline }) => (inline ? "inline-grid" : "grid")};
  ${({ cols }) => cols && `grid-template-columns: ${buildColumns(cols)}`};
  ${({ rows }) => rows && `grid-template-rows: ${buildRows(rows)}`};
  ${({ flow }) => flow && `grid-auto-flow: ${flow}`};
  ${({ gap }) => gap && `grid-gap: ${px(gap)}`};
  ${({ justify }) => justify && `justify-content: ${justify}`};
  ${({ align }) => align && `align-items: ${align}`};
  ${margin};
`;

Grid.defaultProps = {
  cols: 12,
  gap: 24
};

export default Grid;
