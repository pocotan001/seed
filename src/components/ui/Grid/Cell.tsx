import { CSSProperties } from "react";
import styled, { css } from "styled-components";
import media, { IMediaKey, mediaKeys } from "~/components/styles/mixins/media";

interface ICellProps
  extends ICellStyleProps,
    Partial<Record<IMediaKey, ICellStyleProps>> {
  children: React.ReactNode;
}

interface ICellStyleProps {
  col?: number;
  row?: number;
  justify?: CSSProperties["justifySelf"];
  align?: CSSProperties["alignSelf"];
  order?: CSSProperties["order"];
}

const styles = ({ col, row, justify, align, order }: ICellStyleProps) => css`
  ${col && `grid-column-end: span ${col}`};
  ${row && `grid-row-end: span ${row}`};
  ${justify && `justify-self: ${justify}`};
  ${align && `align-self: ${align}`};
  ${order && `order: ${order}`};
`;

const composeMediaStyles = (props: ICellProps) =>
  mediaKeys
    .map(key => props[key] && media[key]`${styles(props[key]!)}`)
    .filter(Boolean);

const Cell = styled<ICellProps, "div">("div")`
  ${styles};
  ${composeMediaStyles};
`;

Cell.defaultProps = {
  col: 1
};

export default Cell;
