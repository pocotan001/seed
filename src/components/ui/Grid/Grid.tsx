import { CSSProperties } from "react";
import styled from "styled-components";
import margin, { IMarginProps } from "~/components/styles/extends/margin";
import media, { IMediaKey, mediaKeys } from "~/components/styles/mixins/media";
import { px } from "~/utils";

interface IGridProps
  extends IGridStyleProps,
    Partial<Record<IMediaKey, IGridStyleProps>>,
    IMarginProps {
  children: React.ReactNode;
}

interface IGridStyleProps {
  cols?: number | CSSProperties["gridTemplateColumns"];
  rows?: number | CSSProperties["gridTemplateRows"];
  flow?: CSSProperties["gridAutoFlow"];
  gap?: CSSProperties["gridGap"];
  justify?: CSSProperties["justifyContent"];
  align?: CSSProperties["alignContent"];
}

const buildColumns = (v: number | string) =>
  typeof v === "number" ? `repeat(${v}, 1fr)` : v;
const buildRows = buildColumns;

const buildStyles = ({
  cols,
  rows,
  flow,
  gap,
  justify,
  align
}: IGridStyleProps) => `
  ${cols && `grid-template-columns: ${buildColumns(cols)}`};
  ${rows && `grid-template-rows: ${buildRows(rows)}`};
  ${flow && `grid-auto-flow: ${flow}`};
  ${gap && `grid-gap: ${px(gap)}`};
  ${justify && `justify-content: ${justify}`};
  ${align && `align-content: ${align}`};
`;

const composeMediaStyles = (props: IGridProps) =>
  mediaKeys
    .map(key => props[key] && media[key]`${buildStyles(props[key]!)}`)
    .filter(Boolean);

/**
 * @example
 * // Different gaps at different breakpoints
 * <Grid gap={30} tablet={{ gap: 20 }} phone={{ gap: 10 }}>
 *   ...
 * </Grid>
 */
const Grid = styled<IGridProps, "div">("div")`
  display: grid;
  ${buildStyles};
  ${composeMediaStyles};
  ${margin};
`;

Grid.defaultProps = {
  cols: 12,
  gap: 24
};

export default Grid;
