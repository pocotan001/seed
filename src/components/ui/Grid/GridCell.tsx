import { CSSProperties } from "react";
import styled from "~/components/styles/themedStyledComponents";

interface IGridCellProps {
  children: React.ReactNode;
  col?: number;
  row?: number;
  justify?: CSSProperties["justifySelf"];
  align?: CSSProperties["alignSelf"];
}

const GridCell = styled<IGridCellProps, "div">("div")`
  ${({ col }) => col && `grid-column-end: span ${col}`};
  ${({ row }) => row && `grid-row-end: span ${row}`};
  ${({ justify }) => justify && `justify-self: ${justify}`};
  ${({ align }) => align && `align-self: ${align}`};
`;

GridCell.defaultProps = {
  col: 1
};

export default GridCell;
