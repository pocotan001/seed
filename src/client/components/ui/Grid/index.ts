import Cell from "./Cell";
import _Grid from "./Grid";

type Grid = typeof _Grid & {
  Cell: typeof Cell;
};

const Grid: Grid = _Grid as Grid;

Grid.Cell = Cell;

export default Grid;
