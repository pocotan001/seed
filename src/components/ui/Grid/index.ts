import Cell from "./Cell";
import _Grid from "./Grid";

const Grid: typeof _Grid & {
  Cell: typeof Cell;
} = _Grid as any;

Grid.Cell = Cell;

export default Grid;
