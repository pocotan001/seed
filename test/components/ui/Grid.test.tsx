import * as React from "react";
import * as renderer from "react-test-renderer";
import Grid from "~/components/ui/Grid";

describe("<Grid>", () => {
  it("should render correctly", () => {
    const tree = renderer
      .create(
        <Grid>
          <Grid.Cell>A</Grid.Cell>
          <Grid.Cell>B</Grid.Cell>
        </Grid>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
