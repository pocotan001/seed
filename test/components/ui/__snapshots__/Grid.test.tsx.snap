// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid> should render correctly 1`] = `
<div
  className="Grid-pvesk6-0 dRvszr"
  cols={12}
>
  <div
    className="Cell-f9knfu-0 bvKviF"
  >
    A
  </div>
  <div
    className="Cell-f9knfu-0 bvKviF"
  >
    B
  </div>
</div>
`;
